misc/python/materialize/checks/all_checks/ssh.py

```diff
@@ -0,0 +1,128 @@
+# Copyright Materialize, Inc. and contributors. All rights reserved.
+#
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+from textwrap import dedent
+
+from materialize.checks.actions import Testdrive
+from materialize.checks.checks import Check
+from materialize.checks.common import KAFKA_SCHEMA_WITH_SINGLE_STRING_FIELD
+
+
+def schemas() -> str:
+    return dedent(KAFKA_SCHEMA_WITH_SINGLE_STRING_FIELD)
+
+
+class SshPg(Check):
+    """"""
+
+    def initialize(self) -> Testdrive:
+        return Testdrive(
+            schemas()
+            + dedent(
+                """
+                > CREATE SECRET pgpass AS 'postgres'
+
+                > CREATE CONNECTION pg TO POSTGRES (
+                  HOST postgres,
+                  DATABASE postgres,
+                  USER postgres,
+                  PASSWORD SECRET pgpass,
+                  SSL MODE require,
+                  SSH TUNNEL thancred);
+
+                $ postgres-execute connection=postgres://postgres:postgres@postgres
+                ALTER USER postgres WITH replication;
+                CREATE TABLE t_ssh (f1 INTEGER);
+                ALTER TABLE t_ssh REPLICA IDENTITY FULL;
+                INSERT INTO t_ssh VALUES (1), (2), (3), (4), (5);
+                CREATE PUBLICATION mz_source_ssh FOR ALL TABLES;
+
+                > CREATE SOURCE mz_source_ssh
+                  FROM POSTGRES CONNECTION pg
+                  (PUBLICATION 'mz_source_ssh')
+                  FOR TABLES (t_ssh);
+                """
+            )
+        )
+
+    def manipulate(self) -> list[Testdrive]:
+        return [
+            Testdrive(schemas() + dedent(s))
+            for s in [
+                """
+                $ postgres-execute connection=postgres://postgres:postgres@postgres
+                INSERT INTO t_ssh VALUES (6), (7), (8), (9), (10);
```

Comment:
I think to cover all bases you need:
- in manipulate #1, to create a new source here, based on the existing tunnel. Create a new tunnel as well
- in manipulate #2, create new sources based on existing tunnels. Drop a tunnel too.

Reply:
I tried that first, but the problem was that only testdrive fragments can be run. We'd need to generalize that to support adding the SSH key I believe:
```python
    public_key = c.sql_query(
        """
        select public_key_1 from mz_ssh_tunnel_connections ssh \
        join mz_connections c on c.id = ssh.id
        where c.name = 'thancred';
        """
    )[0][0]
    c.exec(
        "ssh-bastion-host",
        "bash",
        "-c",
        f"echo '{public_key}' > /etc/authorized_keys/mz",
    )
```
I can imagine 3 ways to make this work:
1. Make ssh-bastion work without the key
2. Add support for this in testdrive via `$ ssh-tunnel-authorize-key`
3. Make initialize/manipulate/validate return a lambda which takes a composition instead of just a Testdrive object

I prefer 1 > 2 > 3, will try in that order to get it working.

Reply:
I couldn't find a good way for 1 and 2. @guswynn Any idea? Can we tell the ssh-bastion to accept any key or Mz to always use the same key for each connection?

Reply:
@def- I think sshd is going to make it very annoying to configure it this way; for 2 we could add support but it would be a different codepath than the normal one when creating the ssh connection

Reply:
Then the current approach or 3 are best. I'll check how difficult 3 is.